sr5824-msr9732-sm9453/new_project.md

```diff
@@ -0,0 +1,148 @@
+evaluate the prospects of AIOps in real world usage.
+
+# Proposed Solution
+
+To get the most value out of the AIOps system, it should be deployed as an independent platform that 
+ingests data from all IT monitoring sources, and acts as a central system of engagement.
+
+Such a platform must be powered by three major types of algorithms that fully automate and 
+streamline three key dimensions of IT operations monitoring:
+
+1. Detection - Machine Learning models trained on past dataset, which would detect any aberration in the
+               metrics and would raise a flag. This would be the entry point of the system.
+   
+2. Triaging - Identifying root causes of problems and recurring issues, so that you can take action 
+              on what has been discovered.
+   
+3. Remediate - Take an appropriate action to avoid/tackle the problem at hand.
+
+These steps occur in a sequential manner, i.e, Once we detect an incident, we triage it and then remediate it.
+
+As part of this project, we are going to deploy the AIOps lifecycle. The way we plan to do this is
+as follows:
+
+1. Detection:
+    We will have a detection API (Flask application) being served from CloudLab. We will have pre-trained 
+    models that would sit on the local file system (we would just pickle them and store them. We are not 
+    making it more complex as our main goal is not to have the complete system. Our main goal is to evaluate
+    all the different techniques).
+   
+2. Triage:
+    This is the most difficult part to accomplish. We do not plan to come up with a complete RCA. We are 
+    going to mock this part with rules, to showcase how all the different parts are stitched together. We
+    are going to use StackStorm[7] for this. StackStorm would have sensors, which would poll the data, call 
+    our detection API, based on the Anomaly Scores, match a corresponding rule and take an appropriate action.
+   
+3. Remediate:
+    This part contains the set of actions that we would want to take. These actions can be as simple as
+    restarting a container. This would again be part of the StackStorm ecosystems. Actions are scripts
+    which would be called by StackStorm when a particular rule is matched against it.
+   
+The reason we want to come-up with the entire AIOps Cycle is to showcase how different detecting algorithms
+can have adverse effects in real world. For example: if we miss an incident, there would not be any rules
+matched and thus there would not be any action taken on it.
+
+The main part of our project is going to be evaluating different Unsupervised Anomaly Detection Algorithms. 
+
+We plan to serve the following models using the API:
+
+1. Statistical:\
+    a. Autoregression (AR)\
+    b. Moving Average (MA)\
+    c. ARIMA\
+    d. SARIMA\
+    e. SARIMAX
+   
+2. Machine Learning Based Models:\
+    a. Isolation Forest (ISF)
+   
+3. Neural Nets:\
+    a. LSTM's\
+    b. HTM
```

Comment:
How do you plan to train these?

Reply:
LSTM's and HTM's can be a little tricky. They are essentially made to work on supervised data. The way we plan to achieve this: For every timestamp t, the label for it would be the value t-1. For example, if CPU utilization at t-1 was 30 and at t it is 40, then the vector would somewhat look like <40 30>.

Why does this work? - By doing this, we are just making the future value of CPU utilization depend on the previous one. In this case, if we start seeing a gradual increase in CPU Utilization, we can predict that it will lead to an incident (given the model has seen similar patterns in the past). This approach is more counterintuitive, and we suppose this would give the model context of its past (the way LSTM's should work).

Reply:
I am not sure about this, but it seems like an intriguing idea so worth trying.

Reply:
Yes. Even we want to evaluate and see how it turns out. We can try this approach with more lookaheads/lookbacks too. Like <t, t-3, t-2, t-1> or vice-versa. It would be nice to see how does it turn out.